pdg/src/snapshots/c2rust_pdg__tests__analysis_test_pdg_snapshot_debug.snap

```diff
@@ -963,6 +959,47 @@ g {
 }
 nodes_that_need_write = [2, 1, 0]
 
-num_graphs = 63
-num_nodes = 709
+g {
+	n[0]:  &_3        _    => _   @ bb2[2]:  fn test_addr_taken_loop;  _20 = &raw mut _3;
+	n[1]:  addr.store n[0] => _   @ bb2[1]:  fn test_addr_taken_loop;  _3 = const 2_i32;
+	n[2]:  copy       n[0] => _4  @ bb5[0]:  fn test_addr_taken_loop;  _4 = &(*_20);
+	n[3]:  addr.load  n[2] => _   @ bb14[3]: fn test_addr_taken_loop;  _18 = (*_4);
+	n[4]:  copy       n[0] => _20 @ bb2[2]:  fn test_addr_taken_loop;  _20 = &raw mut _3;
+	n[5]:  addr.store n[4] => _   @ bb2[1]:  fn test_addr_taken_loop;  _3 = const 2_i32;
+	n[6]:  copy       n[4] => _4  @ bb5[0]:  fn test_addr_taken_loop;  _4 = &(*_20);
+	n[7]:  addr.load  n[6] => _   @ bb14[3]: fn test_addr_taken_loop;  _18 = (*_4);
+	n[8]:  copy       n[4] => _20 @ bb2[2]:  fn test_addr_taken_loop;  _20 = &raw mut _3;
+	n[9]:  addr.store n[8] => _   @ bb2[1]:  fn test_addr_taken_loop;  _3 = const 2_i32;
+	n[10]: copy       n[8] => _4  @ bb5[0]:  fn test_addr_taken_loop;  _4 = &(*_20);
+}
+nodes_that_need_write = [9, 8, 5, 4, 1, 0]
+
+g {
+	n[0]: &_2        _    => _  @ bb1[1]: fn test_addr_taken_cond;  _6 = &raw mut _2;
+	n[1]: addr.store n[0] => _  @ bb1[0]: fn test_addr_taken_cond;  _2 = const 1_i32;
+	n[2]: copy       n[0] => _4 @ bb4[3]: fn test_addr_taken_cond;  _4 = &(*_6);
+	n[3]: addr.load  n[0] => _  @ bb4[6]: fn test_addr_taken_cond;  _5 = (*_6);
+	n[4]: copy       n[0] => _6 @ bb3[1]: fn test_addr_taken_cond;  _6 = &raw mut _2;
```

Comment:
It seems a bit odd to claim that `_6` in the second call is initialized by copying from `_6` in the first call, given none of the first call's locals exist any more.  Though I'm not sure if this will actually cause us any problems.

However, when I tried adding a test case to break this, I got some really odd results.  Here's the function I added to `pointers.rs`:

```Rust
#[no_mangle]
pub fn test_addr_taken_rec(n: usize) {
    if n == 0 {
        return;
    }

    let mut x = 1;
    let y = &mut x;
    test_addr_taken_rec(n - 1);
    *y = 2;
}
```

I got one graph that looks like this, which is fine:

```
g {
	n[0]: &_6        _    => _  @ bb6[4]: fn test_addr_taken_rec;  _11 = &raw mut _6;
	n[1]: addr.store n[0] => _  @ bb6[3]: fn test_addr_taken_rec;  _6 = const 1_i32;
	n[2]: copy       n[0] => _7 @ bb6[7]: fn test_addr_taken_rec;  _7 = &mut (*_11);
	n[3]: addr.store n[2] => _  @ bb7[2]: fn test_addr_taken_rec;  (*_7) = const 2_i32;
}
```

But I also got another graph that looks like this:

```
g {
	n[0]:  &_1         _    => _    @ bb0[2]:  fn test_unique_ref;      _8 = &raw mut _1;
	n[1]:  addr.store  n[0] => _    @ bb0[1]:  fn test_unique_ref;      _1 = const 10_i32;
	n[2]:  copy        n[0] => _4   @ bb0[9]:  fn test_unique_ref;      _4 = &mut (*_8);
	n[3]:  copy        n[2] => _3   @ bb0[10]: fn test_unique_ref;      _3 = &raw mut (*_4);
	n[4]:  copy        n[0] => _7   @ bb0[18]: fn test_unique_ref;      _7 = &mut (*_8);
	n[5]:  copy        n[4] => _6   @ bb0[19]: fn test_unique_ref;      _6 = &raw mut (*_7);
	n[6]:  value.store n[5] => _5.* @ bb0[20]: fn test_unique_ref;      (*_5) = move _6;
	n[7]:  copy        n[5] => _11  @ bb6[4]:  fn test_addr_taken_rec;  _11 = &raw mut _6;
	n[8]:  addr.store  n[7] => _    @ bb6[3]:  fn test_addr_taken_rec;  _6 = const 1_i32;
	n[9]:  copy        n[7] => _7   @ bb6[7]:  fn test_addr_taken_rec;  _7 = &mut (*_11);
	n[10]: addr.store  n[9] => _    @ bb7[2]:  fn test_addr_taken_rec;  (*_7) = const 2_i32;
}
```

This is clearly bogus - no pointers flow from `test_unique_ref` to `test_addr_taken_rec` or vice versa.  My guess is that `test_addr_taken_local`'s `_6` (in the recursive call) happened to land at the same address on the stack as `test_unique_ref`'s `_1`, and the latter was still considered live (wrongly), so the initialization of `_11 = &raw mut _6` was treated as a copy of `test_unique_ref`'s `_1`.

I think the instrumentation and/or PDG builder needs to mark addr-taken locals as dead at some point (the same way heap allocations are marked as dead when passed to `free`, so no more nodes can be added to their graphs).  One option is to do this when returning from the function, but that might be a pain.  Another option might be to instrument `StorageLive` for addr-taken locals to mark all objects that overlap the local as dead.

Reply:
I'm aware this is a bug and am working on it